pkg/controller/issueapi/gen_code.go

```diff
@@ -0,0 +1,185 @@
+)
+
+const (
+	// all lowercase characters plus 0-9
+	charset = "abcdefghijklmnopqrstuvwxyz0123456789"
+)
+
+func (c *Controller) IssueCode(ctx context.Context, vCode *database.VerificationCode, realm *database.Realm) *IssueResult {
+	logger := logging.FromContext(ctx).Named("issueapi.IssueCode")
+
+	// If we got this far, we're about to issue a code - take from the limiter
+	// to ensure this is permitted.
+	if realm.AbusePreventionEnabled {
+		key, err := realm.QuotaKey(c.config.GetRateLimitConfig().HMACKey)
+		if err != nil {
+			return &IssueResult{
+				obsResult:   observability.ResultError("FAILED_TO_GENERATE_HMAC"),
+				HTTPCode:    http.StatusInternalServerError,
+				ErrorReturn: api.Error(err).WithCode(api.ErrInternal),
+			}
+		}
+
+		if limit, _, reset, ok, err := c.limiter.Take(ctx, key); err != nil {
+			logger.Errorw("failed to take from limiter", "error", err)
+			return &IssueResult{
+				obsResult:   observability.ResultError("FAILED_TO_TAKE_FROM_LIMITER"),
+				HTTPCode:    http.StatusInternalServerError,
+				ErrorReturn: api.Errorf("failed to issue code, please try again in a few seconds").WithCode(api.ErrInternal),
+			}
+		} else if !ok {
+			logger.Warnw("realm has exceeded daily quota",
+				"realm", realm.ID,
+				"limit", limit,
+				"reset", reset)
+
+			if c.config.GetEnforceRealmQuotas() {
+				return &IssueResult{
+					obsResult:   observability.ResultError("QUOTA_EXCEEDED"),
+					HTTPCode:    http.StatusTooManyRequests,
+					ErrorReturn: api.Errorf("exceeded daily realm quota configured from abuse prevention, please contact a realm administrator").WithCode(api.ErrQuotaExceeded),
+				}
+			}
+		}
+		stats.Record(ctx, mRealmTokenUsed.M(1))
+	}
+
+	if err := c.CommitCode(ctx, vCode, realm, c.config.GetCollisionRetryCount()); err != nil {
+		logger.Errorw("failed to issue code", "error", err)
+		return &IssueResult{
+			obsResult:   observability.ResultError("FAILED_TO_ISSUE_CODE"),
+			HTTPCode:    http.StatusInternalServerError,
+			ErrorReturn: api.Errorf("failed to generate otp code, please try again").WithCode(api.ErrInternal),
+		}
+	}
+
+	return &IssueResult{
+		VerCode:   vCode,
+		HTTPCode:  http.StatusOK,
+		obsResult: observability.ResultOK(),
+	}
+}
+
+// CommitCode will generate a verification code and save it to the database, based on
+// the paremters provided. It returns the short code, long code, a UUID for
+// accessing the code, and any errors.
+func (c *Controller) CommitCode(ctx context.Context, vCode *database.VerificationCode, realm *database.Realm, retryCount uint) error {
+	logger := logging.FromContext(ctx)
+	var err error
+
+	b, err := retry.NewConstant(50 * time.Millisecond)
+	if err != nil {
+		return fmt.Errorf("failed to create backoff: %w", err)
+	}
+
+	retry.Do(ctx, retry.WithMaxRetries(uint64(retryCount), b), func(ctx context.Context) error {
+		var code string
+		code, err = GenerateCode(realm.CodeLength)
+		if err != nil {
+			logger.Errorf("code generation error: %v", err)
```

Comment:
Can we remove these log messages or let the error propagate up? Else, we should use `logger.Errorw`

Reply:
yea looks like there's an Errorw one up from this. Let's leave it at that.